File: python/problem12.py

import math

def factor(number):
    factors = [number,1]

    root = math.floor(number ** 0.5)
    if root * root == number:
        factors.append(root)

    n = 2
    while n < root:
        if number % n == 0:
            factors.append(n)
            factors.append(number / n)

        n += 1

    return factors

def triangle():
    yield 1

    t = 1
    n = 2

    while True:
        t += n
        yield t

        n += 1

triangles = triangle()

for t in triangles:
    f = factor(t)

    if len(f) > 500:
        print t
        break

File: cpp/problem12.cpp
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

using namespace std;

#define uint unsigned int

int main(int argc, char* argv[])
{
	uint n = 2;
	uint triangle = 1;
	uint root;
	uint factor;
	uint count;

	do
	{
		triangle += n++;
		count = 2;

		root = floor(sqrt(triangle));
		if (root * root == triangle)
		{
			count++;
		}

		for (factor = 2; factor < root; factor++)
		{
			if (triangle % factor == 0)
			{
				count += 2;
			}
		}
	}
	while (count <= 500);

	printf("%u\n", triangle);

	return 0;
}

